File: keyboard.cpp
#include "keyboard.h"
#include <GL/glfw.h>
#include <stdio.h>

Keyboard::Keyboard() 
{
	resetfunction ();
}
Keyboard::~Keyboard() 
{
}
void Keyboard::updateKeyboardStatus() 
{ 
	resetfunction () ;
	bool foundKey=false;
	//0~9
	for(int key=33;key<=126;key++)
	{
		if(glfwGetKey(key)==1)
		{
			keyChar = key;
			keyStatus=0;
			foundKey=true;
			break;
		}
	}
	if(foundKey==false)
		return;

	//shift
	if(glfwGetKey(GLFW_KEY_LSHIFT)==1 || glfwGetKey(GLFW_KEY_RSHIFT)==1)
		shiftKey = 1;
	else
		shiftKey = 0;
	//control
	if(glfwGetKey(GLFW_KEY_LCTRL)==1 || glfwGetKey(GLFW_KEY_RCTRL)==1)
		controlKey = 1;
	else
		controlKey = 0;
	
	//capital character
	if(keyChar>='A' && keyChar<='Z')
		if(controlKey==1)
			keyChar=keyChar;//keep cap character
		else
		{
			if(shiftKey==0)
				keyChar+=32;
			else
				keyChar=keyChar;
		}


	printf("keyChar %c \n", keyChar);	
	//printf("shiftKey %d \n", shiftKey); 
	//printf("controlKey %d \n", controlKey); 
}
void Keyboard::setPropertes (int t_controlKey, int t_shiftKey, char t_keyChar, int t_keyStatus) {
	controlKey = controlKey;
	shiftKey = shiftKey;
	keyChar = t_keyChar;
	keyStatus = keyStatus;
}
void Keyboard::onKeyDown(int t_controlKey, int t_shiftKey, int t_keyChar) {
	setPropertes(t_controlKey, t_shiftKey, t_keyChar, 0);
}
void Keyboard::onKeyUp(int t_controlKey, int t_shiftKey, int t_keyChar) {
	setPropertes(t_controlKey, t_shiftKey, t_keyChar, 1);
}
void Keyboard::onKeyPress(int t_controlKey, int t_shiftKey, int t_keyChar) {
	setPropertes(t_controlKey, t_shiftKey, t_keyChar, 2);
}
void Keyboard::resetfunction () {
	keyChar = -1; //ASCII code
	keyStatus = -1;//up 1, down 0
	controlKey = -1; //up 1, down 0
	shiftKey = -1; //up 1, down 0
}

File: keyboard.js
function keyboardClass() {
    var keyboard = {};
    //keyboard public member variables
    keyboard.char = null; //ASCII code
    keyboard.keyStatus = -1;//up 1, down 0
    keyboard.controlKey = -1; //up 1, down 0
    keyboard.shiftKey = -1; //up 1, down 0
    //keyboard public member methods
    keyboard.setPropertes = function (controlKey, shiftKey, char, keyStatus) {
        keyboard.controlKey = controlKey;
        keyboard.shiftKey = shiftKey;
        keyboard.char = char;
        keyboard.keyStatus = keyStatus;
    }
    keyboard.keyboardEvent = function (handler, event) {
        var controlKey, shiftKey, controlKey;
        window.event.ctrlKey ? controlKey = 0 : controlKey = 1;
        window.event.shiftKey ? shiftKey = 0 : shiftKey = 1;
        char = event.char;
        handler(controlKey, shiftKey, char);
    }
    keyboard.onKeyDown = function (controlKey, shiftKey, char) {
        keyboard.setPropertes(controlKey, shiftKey, char, 0);
    }
    keyboard.onKeyUp = function (controlKey, shiftKey, char) {
        keyboard.setPropertes(controlKey, shiftKey, char, 1);
    }
    keyboard.onKeyPress = function (controlKey, shiftKey, char) {
        keyboard.setPropertes(controlKey, shiftKey, char, 2);
    }
    keyboard.reset = function () {
        keyboard.char = null; //ASCII code
        keyboard.keyStatus = -1;//up 1, down 0
        keyboard.controlKey = -1; //up 1, down 0
        keyboard.shiftKey = -1; //up 1, down 0
    }
    return keyboard;
}